docs/proposals/resource-aggregation-proxy/README.md

```diff
@@ -0,0 +1,455 @@
+
+// ProxyOptions define a flag for proxy that do not have actual resources.
+type ProxyOptions struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Path is the part of URLs used for the current proxy request.
+	// For example, the whole request URL is
+	// http://localhost/apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy/api/v1/nodes
+	// Path is api/v1/nodes
+	// +optional
+	Path string `json:"path,omitempty" protobuf:"bytes,1,opt,name=path"`
+}
+```
+
+### Example
+
+Users create custom `GlobalResource` as below:
+
+```yaml
+apiVersion: proxy.karmada.io/v1alpha1
+kind:  GlobalResource
+metadata:
+  name: proxy-sample
+spec:
+  targetCluster:
+    clusterNames:
+    - member1
+    - member2
+    - member3
+  resourceSelectors:
+    - apiVersion: v1
+      kind: Pod
+    - apiVersion: v1
+      kind: Node
+```
+
+
+
+## How to Merge Result
+
+- Get request: return first result.
+- List request: concat the results, and return to client as a single list.
+- Watch request: return result one by one.
+
+## Resource Version
+
+Kubernetes uses ResourceVersion to uniquely identify every version of a resource. We merge all clusters' ResourceVersion into one JSON struct, `base64.RawURLEncoding` it, and get the final string.
+
+```
+base64.RawURLEncoding({"cluster1":"1234","cluster2":"5678"}) = "eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ"
+```
+
+## Example
+
+There are two member clusters: cluster1 and cluster2.
+
+|                 | cluster1         | cluster2         |
+| --------------- | ---------------- | ---------------- |
+| resourceVersion | 1234             | 5678             |
+| pods            | nginx-1, nginx-2 | nginx-3, nginx-4 |
+
+**List Pod**
+
+Situation: Client wants to list all pods in cluster1 and cluster2.
+
+
+
+1\. Client sends request with prefixed url.
+
+```
+GET /apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy/api/v1/namespaces/default/pods
+```
+
+2\. Proxy query clusters1's cache, get nginx-1, nginx-2; query cluster2's cache, get nginx-3, nginx-4.
+
+3\. Proxy merges pods' list and returns.
+
+```yaml
+apiVersion: v1
+kind: List
+metadata:
+  # eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ is result of 
+  # base64.RawURLEncoding("{"cluster1":"1234","cluster2":"5678"}")
+  resourceVersion: eyJjbHVzdGVyMSI6IjEyMzQiLCJjbHVzdGVyMiI6IjU2NzgifQ
+items:
+- apiVersion: v1
+  kind: Pod
+  metadata:
+    name: nginx-1
+  ...
+- apiVersion: v1
+  kind: Pod
+  metadata:
+    name: nginx-2
+  ...
+- apiVersion: v1
+  kind: Pod
+  metadata:
+    name: nginx-3
+  ...
+- apiVersion: v1
+  kind: Pod
+  metadata:
+    name: nginx-4
+  ...
+```
+
+
+
+**Get Pod**
+
+Situation: Client wants to get Pod by name "nginx-2".
+
+
+
+1\. Client sends request with prefixed url.
+
+```
+GET /apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy/api/v1/namespaces/default/pods/nginx-2
+```
+
+2\. Proxy queries all caches, and finds out the required Pod is in cluster1's cache.
```

Comment:
I know we assume that each pod is unique, but what if two pods with the same name, what's the behavior here?
returns one by random or panic?

Reply:
Good question. I prefer to return the first one (order by the cluster name). For these reasons:
1. panic is two serious.
2. the returns is stable for client.

Reply:
I agree.
I think ensuring each pod is unique is the biggest challenge for this feature. Do you have any idea to solve it?